osmoutils/osmocli/parsers.go

```diff
@@ -0,0 +1,282 @@
+package osmocli
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/gogo/protobuf/proto"
+	"github.com/spf13/pflag"
+
+	"github.com/osmosis-labs/osmosis/v13/osmoutils"
+)
+
+// Parses arguments 1-1 from args
+// makes an exception, where it allows Pagination to come from flags.
+func ParseFieldsFromFlagsAndArgs[reqP proto.Message](flagAdvice FlagAdvice, flags *pflag.FlagSet, args []string) (reqP, error) {
+	req := osmoutils.MakeNew[reqP]()
+	v := reflect.ValueOf(req).Elem()
+	t := v.Type()
+
+	argIndexOffset := 0
+	// Iterate over the fields in the struct
+	for i := 0; i < t.NumField(); i++ {
+		arg := ""
+		if len(args) > i+argIndexOffset {
+			arg = args[i+argIndexOffset]
+		}
+		usedArg, err := ParseField(v, t, i, arg, flagAdvice, flags)
+		if err != nil {
+			return req, err
+		}
+		if !usedArg {
+			argIndexOffset -= 1
+		}
+	}
+	return req, nil
+}
+
+func ParseNumFields[reqP proto.Message]() int {
+	req := osmoutils.MakeNew[reqP]()
+	v := reflect.ValueOf(req).Elem()
+	t := v.Type()
+	return t.NumField()
+}
+
+func ParseExpectedFnName[reqP proto.Message]() string {
+	req := osmoutils.MakeNew[reqP]()
+	v := reflect.ValueOf(req).Elem()
+	s := v.Type().String()
+	// handle some non-std queries
+	var prefixTrimmed string
+	if strings.Contains(s, "Query") {
+		prefixTrimmed = strings.Split(s, "Query")[1]
+	} else {
+		prefixTrimmed = strings.Split(s, ".")[1]
+	}
+	suffixTrimmed := strings.TrimSuffix(prefixTrimmed, "Request")
+	return suffixTrimmed
+}
+
+func ParseHasPagination[reqP proto.Message]() bool {
+	req := osmoutils.MakeNew[reqP]()
+	t := reflect.ValueOf(req).Elem().Type()
+	for i := 0; i < t.NumField(); i++ {
+		fType := t.Field(i)
+		if fType.Type.String() == paginationType {
+			return true
+		}
+	}
+	return false
+}
+
+const paginationType = "*query.PageRequest"
+
+type FlagAdvice struct {
```

Comment:
nit: suggest moving these up to the top to follow the style conventions in the rest of the repo where the order is 1. types, 2. consts 3. function definitions